test/apps/test_hooks.c

```diff
@@ -0,0 +1,227 @@
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    return UCS_OK;
+}
+
+static uint64_t get_time_msec(void)
+{
+    struct timeval tv;
+
+    gettimeofday(&tv, NULL);
+    return (uint64_t)tv.tv_sec * 1000 + (tv.tv_usec / 1000);
+}
+
+static void *thread_handler(void *arg)
+{
+    context_t *ctx    = arg;
+    uint64_t deadline = get_time_msec() + ctx->thread_timeout_msec;
+
+    /*
+     * Many threads racing for read lock are preventing ucm to finish testing
+     * the installed hooks, so we make sure to stop the threads after
+     * reasonable time.
+     */
+    do {
+        if (mmap_call(0) != MAP_FAILED) {
+            printf("failed to return mmap_call() failure\n");
+            exit(EXIT_FAILURE);
+        }
+    } while (!ctx->stop &&
+             ((ctx->thread_timeout_msec == 0) || (get_time_msec() < deadline)));
+
+    return NULL;
+}
+
+static void ctx_stop_threads(context_t *ctx)
+{
+    int i;
+
+    printf("stopping %d threads\n", ctx->thread_count);
+
+    ctx->stop = 1;
+    for (i = 0; i < ctx->thread_count; i++) {
+        pthread_join(ctx->threads[i], NULL);
+    }
+}
+
+static ucs_status_t ctx_start_threads(context_t *ctx)
+{
+    int i, ret;
+
+    for (i = 0; i < ctx->thread_count; i++) {
+        ret = pthread_create(&ctx->threads[i], NULL, thread_handler, ctx);
+        if (ret < 0) {
+            printf("pthread_create(thread #%d) failed: %m\n", i);
+            goto err;
+        }
+    }
+
+    printf("started %d threads\n", ctx->thread_count);
+    return UCS_OK;
+
+err:
+    ctx->thread_count = i;
+    ctx_stop_threads(ctx);
+    return UCS_ERR_NO_RESOURCE;
+}
+
+static void ctx_free(context_t *ctx)
+{
+    if (ctx->dl_handle != NULL) {
+        dlclose(ctx->dl_handle);
+    }
+    free(ctx->threads);
+}
+
+static void usage(const char *argv0)
+{
+    printf("Usage: %s [options]\n", argv0);
+    printf("Options:\n");
+    printf("  -n :  Number of threads racing with function patching (default: "
+           "%d)\n",
+           DEFAULT_THREAD_COUNT);
+    printf("  -d :  Disable run time timeouts (thread: %d sec, overall: %d "
+           "sec)\n",
+           DEFAULT_THREAD_TIMEOUT, DEFAULT_TIMEOUT);
+    printf("  -h :  Display help message\n");
+    printf("\n");
+}
+
+int main(int argc, char **argv)
+{
+    int thread_count   = DEFAULT_THREAD_COUNT;
+    int thread_timeout = DEFAULT_THREAD_TIMEOUT;
+    int timeout        = DEFAULT_TIMEOUT;
+    int delay_usec     = 200 * 1000;
+    int c, ret;
+    context_t ctx;
+    ucs_status_t status;
+
+    while ((c = getopt(argc, argv, "n:dh")) != -1) {
+        switch (c) {
+        case 'n':
+            thread_count = atoi(optarg);
+            break;
+        case 'd':
+            timeout        = 0;
+            thread_timeout = 0;
+            break;
+        case 'h':
+        default:
+            usage(argv[0]);
+            return -1;
+        }
+    }
+
+    if (timeout > 0) {
+        alarm(timeout); /* patching could corrupt threads */
+    }
```

Comment:
IMO better run the test with "timeout" command to kill it if it's stuck, rather than relying on internal timers - that could be corrupted. it also allows easier debug of the test when it hangs and simplifies the test by removing arguments parsing code.